### PR TITLE
ACME Certificate Revocation

### PR DESCRIPTION
**Pull request by hslatman**
This PR adds ACME certificate revocation as discussed in #468. 

Basic revocation using either JWK or KID works. Needs error handling improvements and tests.

### CODE REVIEW

File: acme/api/handler.go [hunk 1]
```diff
@@ -100,11 +100,17 @@ func (h *Handler) Route(r api.Router) {
 	r.MethodFunc("GET", getPath(DirectoryLinkType, "{provisionerID}"), h.baseURLFromRequest(h.lookupProvisioner(h.GetDirectory)))
 	r.MethodFunc("HEAD", getPath(DirectoryLinkType, "{provisionerID}"), h.baseURLFromRequest(h.lookupProvisioner(h.GetDirectory)))
 
+	validatingMiddleware := func(next nextHTTP) nextHTTP {
+		return h.baseURLFromRequest(h.lookupProvisioner(h.addNonce(h.addDirLink(h.verifyContentType(h.parseJWS(next))))))
+	}
 	extractPayloadByJWK := func(next nextHTTP) nextHTTP {
-		return h.baseURLFromRequest(h.lookupProvisioner(h.addNonce(h.addDirLink(h.verifyContentType(h.parseJWS(h.validateJWS(h.extractJWK(h.verifyAndExtractJWSPayload(next)))))))))
+		return validatingMiddleware(h.validateJWS(h.extractJWK(h.verifyAndExtractJWSPayload(next))))
```
Comment (dopey) [inline acme/api/handler.go:107]: should ValidateJWS also be part of validateMiddleware?

File: acme/api/revoke.go [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+		return "ip|" + identifier.Value
+	}
+	if identifier.Type == acme.DNS {
+		return "dns|" + identifier.Value
+	}
+	return "unsupported|" + identifier.Value
+}
+
+// extractIdentifiers extracts ACME identifiers from an x509 certificate and
+// creates a map from them. The map ensures that double SANs are deduplicated.
```
Comment (dopey) [inline acme/api/revoke.go:208]: double -> duplicated

File: acme/api/revoke.go [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+		result[identifierKey(identifier)] = identifier
+	}
+	for _, ip := range cert.IPAddresses {
+		identifier := acme.Identifier{
+			Type:  acme.IP,
+			Value: ip.String(),
+		}
+		result[identifierKey(identifier)] = identifier
+	}
+	// TODO(hs): should we include the CommonName or not?
```
Comment (dopey) [inline acme/api/revoke.go:230]: Seems like we can remove this comment.

File: acme/api/handler.go (hunk 1 above)
Comment (hslatman) [reply acme/api/handler.go]: Good point 😅. I have to admit `validatingMiddleware` is not that great of a name, as it does more than just that, but it works.

Fixed in 004fc054d570908479915306d98ca07761ccd491.

File: acme/api/revoke.go (hunk 2 above)
Comment (hslatman) [reply acme/api/revoke.go]: Fixed in 004fc054d570908479915306d98ca07761ccd491

File: acme/api/revoke.go (hunk 3 above)
Comment (hslatman) [reply acme/api/revoke.go]: Fixed in 004fc054d570908479915306d98ca07761ccd491

**Review by dopey — APPROVED** [review]
lgtm

**Pull request merged**